### PR TITLE
JBS-24 : added api for cinder ec2

### PR DESCRIPTION
**Pull request by rjain-iitr**

### CODE REVIEW

File: ec2api/api/cloud.py [hunk 1]
```diff
@@ -851,10 +851,8 @@ def delete_snapshot(self, context, snapshot_id):
             Returns true if the request succeeds.
         """
 
-    @module_and_param_types(snapshot, 'snap_ids', 'strs',
-                            'strs', 'filter')
-    def describe_snapshots(self, context, snapshot_id=None, owner=None,
-                           restorable_by=None, filter=None):
+    @module_and_param_types(snapshot, 'str','int','str')
+    def describe_snapshots(self, context, snapshot_id=None,limit=None, marker=None):
```
Comment (aggarwalchirag) [inline ec2api/api/cloud.py]: No detail handling for snapshots.

File: ec2api/api/volume.py [hunk 2]
```diff
@@ -30,28 +30,37 @@
...
+                  snapshot_id=None,
+                  name=None,description=None):
     cinder = clients.cinder(context)
+    if size is None :
+       size=0
+    if snapshot_id is not None:
+       os_snapshot=cinder.backups.get(snapshot_id)
+       snap_size=os_snapshot.size
+       if snap_size >size :
+           size=snap_size
```
Comment (aggarwalchirag) [inline ec2api/api/volume.py]: size handling is incorrect when a snapshot is provided

File: ec2api/api/volume.py [hunk 3]
```diff
@@ -30,28 +30,37 @@
...
     with common.OnCrashCleaner() as cleaner:
         os_volume = cinder.volumes.create(
-                size, snapshot_id=os_snapshot_id, volume_type=volume_type,
-                availability_zone=availability_zone)
+            size,name=name,description=description)
```
Comment (aggarwalchirag) [inline ec2api/api/volume.py]: creation of volumes from snapshot has not been handled.

File: ec2api/api/volume.py [hunk 4]
```diff
@@ -30,28 +30,37 @@
...
-
-        volume = db_api.add_item(context, 'vol', {'os_id': os_volume.id})
-        cleaner.addCleanup(db_api.delete_item, context, volume['id'])
-        os_volume.update(display_name=volume['id'])
-
-    return _format_volume(context, volume, os_volume, snapshot_id=snapshot_id)
-
+        #volume = db_api.add_item(context, 'vol', {'os_id': os_volume.id})
+        #cleaner.addCleanup(db_api.delete_item, context, volume['id'])
+        if snapshot_id is not None:
```
Comment (aggarwalchirag) [inline ec2api/api/volume.py]: can't volume creation and restore from volume be performed as a single step ?

File: ec2api/api/volume.py [hunk 5]
```diff
@@ -30,28 +30,37 @@
...
+              import time
+              time.sleep(5)
+              os_volume=cinder.restores.restore(backup_id=snapshot_id,volume_id=os_volume.id)
+              return True
```
Comment (aggarwalchirag) [inline ec2api/api/volume.py]: why aren't we calling format here ?

File: ec2api/api/volume.py [hunk 6]
```diff
@@ -92,31 +101,32 @@ def detach_volume(context, volume_id, instance_id=None, device=None,
...
     cinder = clients.cinder(context)
     try:
-        cinder.volumes.delete(volume['os_id'])
+        cinder.volumes.delete(volume_id)
     except cinder_exception.BadRequest:
         # TODO(andrey-mp): raise correct errors for different cases
         raise exception.UnsupportedOperation()
     except cinder_exception.NotFound:
         pass
+    os_volume = cinder.volumes.get(volume_id)
```
Comment (aggarwalchirag) [inline ec2api/api/volume.py]: race condition handling ?

File: ec2api/api/volume.py [hunk 7]
```diff
@@ -30,28 +30,37 @@
...
-        snapshot = ec2utils.get_db_item(context, snapshot_id)
-        os_snapshot_id = snapshot['os_id']
-    else:
-        os_snapshot_id = None
-
+def create_volume(context, size=None,
+                  snapshot_id=None,
+                  name=None,description=None):
     cinder = clients.cinder(context)
+    if size is None :
```
Comment (aggarwalchirag) [inline ec2api/api/volume.py]: if empty volume is to be created, then size is mandatory. Check for the same is missing.

File: ec2api/api/volume.py [hunk 8]
```diff
@@ -172,6 +244,7 @@ def _format_volume(context, volume, os_volume, instances={},
                 context, 'snap', os_volume.snapshot_id, snapshots)
         snapshot_id = snapshot['id']
     ec2_volume['snapshotId'] = snapshot_id
+    #ec2_volume['name'] = os_volume.get('name')
```
Comment (aggarwalchirag) [inline ec2api/api/volume.py]: why do we need to look up db to get the snapshot id ?

File: ec2api/api/volume.py [hunk 9]
```diff
@@ -153,14 +226,13 @@ def _format_volume(context, volume, os_volume, instances={},
...
     ec2_volume = {
-            'volumeId': volume['id'],
+            'volumeId': os_volume.id,
             'status': valid_ec2_api_volume_status_map.get(os_volume.status,
                                                           os_volume.status),
             'size': os_volume.size,
-            'availabilityZone': os_volume.availability_zone,
+            'name': os_volume.name,
+            'description': os_volume.description,
             'createTime': os_volume.created_at,
```
Comment (aggarwalchirag) [inline ec2api/api/volume.py]: the response field is CreatedAt. Also, the case of the fields is not in sync with the document.

File: ec2api/api/volume.py [hunk 10]
```diff
@@ -153,14 +226,13 @@ def _format_volume(context, volume, os_volume, instances={},
         'detaching': 'in-use'}
 
     ec2_volume = {
-            'volumeId': volume['id'],
+            'volumeId': os_volume.id,
             'status': valid_ec2_api_volume_status_map.get(os_volume.status,
```
Comment (aggarwalchirag) [inline ec2api/api/volume.py]: status map is incorrect.

File: ec2api/api/volume.py [hunk 11]
```diff
@@ -138,13 +148,76 @@ def get_os_items(self):
...
+def _format_volume_delete(context, os_volume):
+    valid_ec2_api_volume_status_map = {
+        'attaching': 'in-use',
+        'detaching': 'in-use',
+        'in-use': 'in-use',
+        'deleting': 'deleting',
+        'backingup': 'in-use'}
+
+    ec2_volume = {
+            'status': valid_ec2_api_volume_status_map.get(os_volume.status,
```
Comment (aggarwalchirag) [inline ec2api/api/volume.py]: the status map needs work

File: ec2api/api/snapshot.py [hunk 12]
```diff
@@ -42,39 +42,64 @@ def create_snapshot(context, volume_id, description=None):
...
 def delete_snapshot(context, snapshot_id):
-    snapshot = ec2utils.get_db_item(context, snapshot_id)
+    #snapshot = ec2utils.get_db_item(context, snapshot_id)
     cinder = clients.cinder(context)
     try:
-        cinder.volume_snapshots.delete(snapshot['os_id'])
+        cinder.backups.delete(snapshot_id)
     except cinder_exception.NotFound:
         pass
+    os_snapshot=cinder.backups.get(snapshot_id)
```
Comment (aggarwalchirag) [inline ec2api/api/snapshot.py]: race condition handling ?

File: ec2api/api/volume.py (hunk 6 above)
Comment (rjain-iitr) [inline ec2api/api/volume.py]: cinder.volumes.delete(volume_id) doesnot return an output so need to changes the cinder volume delete api for this.

File: ec2api/api/cloud.py (hunk 1 above)
Comment (rjain-iitr) [inline ec2api/api/cloud.py]: Missed changes in cloud.py

File: ec2api/api/snapshot.py (hunk 12 above)
Comment (rjain-iitr) [inline ec2api/api/snapshot.py]: need to change cinder backups api to return response while deleting volume currently it is not returning any thing.

**Pull request merged**